File: objectModel/CSharp/Microsoft.CommonDataModel.ObjectModel/Cdm/CdmPurposeReference.cs
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License. See License.txt in the project root for license information.

namespace Microsoft.CommonDataModel.ObjectModel.Cdm
{
    using Microsoft.CommonDataModel.ObjectModel.Enums;
    using Microsoft.CommonDataModel.ObjectModel.Utilities;
    using System;

    public class CdmPurposeReference : CdmObjectReferenceBase
    {
        /// <summary>
        /// Constructs a CdmPurposeReference.
        /// </summary>
        /// <param name="ctx">The context.</param>
        /// <param name="purpose">The purpose to reference.</param>
        /// <param name="simpleReference">Whether this reference is a simple reference.</param>
        public CdmPurposeReference(CdmCorpusContext ctx, dynamic purpose, bool simpleReference)
            : base(ctx, (object)purpose, simpleReference)
        {
            this.ObjectType = CdmObjectType.PurposeRef;
        }

        [Obsolete]
        public override CdmObjectType GetObjectType()
        {
            return CdmObjectType.PurposeRef;
        }

        internal override CdmObjectReferenceBase CopyRefObject(ResolveOptions resOpt, dynamic refTo, bool simpleReference, CdmObjectReferenceBase host = null)
        {
            if (host == null)
                return new CdmPurposeReference(this.Ctx, refTo, simpleReference);
            else
                return host.CopyToHost(this.Ctx, refTo, simpleReference);
        }

        [Obsolete("CopyData is deprecated. Please use the Persistence Layer instead.")]
        public override dynamic CopyData(ResolveOptions resOpt, CopyOptions options)
        {
            return CdmObjectBase.CopyData<CdmPurposeReference>(this, resOpt, options);
        }

        internal override bool VisitRef(string pathFrom, VisitCallback preChildren, VisitCallback postChildren)
        {
            return false;
        }
    }
}

File: objectModel/TypeScript/Cdm/CdmPurposeReference.ts
// Copyright (c) Microsoft Corporation. All rights reserved.
// Licensed under the MIT License. See License.txt in the project root for license information.

import {
    CdmCorpusContext,
    CdmObjectReferenceBase,
    cdmObjectType,
    CdmPurposeDefinition,
    resolveOptions,
    VisitCallback
} from '../internal';

export class CdmPurposeReference extends CdmObjectReferenceBase {
    public static get objectType(): cdmObjectType {
        return cdmObjectType.purposeRef;
    }

    constructor(ctx: CdmCorpusContext, purpose: string | CdmPurposeDefinition, simpleReference: boolean) {
        super(ctx, purpose, simpleReference);
        // let bodyCode = () =>
        {
            this.objectType = cdmObjectType.purposeRef;
        }
        // return p.measure(bodyCode);
    }

    public getObjectType(): cdmObjectType {
        // let bodyCode = () =>
        {
            return cdmObjectType.purposeRef;
        }
        // return p.measure(bodyCode);
    }

    /**
     * @internal
     */
    public copyRefObject(resOpt: resolveOptions, refTo: string | CdmPurposeDefinition, simpleReference: boolean, host?: CdmObjectReferenceBase): CdmObjectReferenceBase {
        // let bodyCode = () =>
        {
            if (!host) {
                return new CdmPurposeReference(this.ctx, refTo, simpleReference);
            } else {
                return host.copyToHost(this.ctx, refTo, simpleReference);
            }
        }
        // return p.measure(bodyCode);
    }

    /**
     * @internal
     */
    public visitRef(pathFrom: string, preChildren: VisitCallback, postChildren: VisitCallback): boolean {
        // let bodyCode = () =>
        {
            return false;
        }
        // return p.measure(bodyCode);
    }
}
